packages/assets-controllers/src/multicall.ts

```diff
@@ -416,3 +466,402 @@ export const multicallOrFallback = async (
+      balanceMap[ZERO_ADDRESS] = nativeBalances;
+    }
+  }
+
+  return balanceMap;
+};
+
+/**
+ * Get token balances (both ERC20 and native) for multiple addresses using aggregate3.
+ * This is more efficient than individual balanceOf calls for multiple addresses and tokens.
+ * Native token balances are mapped to the zero address (0x0000000000000000000000000000000000000000).
+ *
+ * @param tokenAddresses - Array of ERC20 token contract addresses, or null/empty array for native balances only
+ * @param userAddresses - Array of user addresses to check balances for
+ * @param chainId - The hexadecimal chain id
+ * @param provider - An ethers rpc provider
+ * @param includeNative - Whether to include native token balances (default: true)
+ * @returns Promise resolving to map of token address to map of user address to balance
+ */
+export const getTokenBalancesForMultipleAddresses = async (
+  tokenAddresses: string[] | null,
+  userAddresses: string[],
+  chainId: Hex,
+  provider: Web3Provider,
+  includeNative = true,
+): Promise<Record<string, Record<string, BN>>> => {
+  // Normalize tokenAddresses to empty array if null
+  const normalizedTokenAddresses = tokenAddresses || [];
+
+  // Return early only if no user addresses or if no tokens and no native balances requested
+  if (
+    userAddresses.length === 0 ||
+    (normalizedTokenAddresses.length === 0 && !includeNative)
+  ) {
+    return {};
+  }
+
+  // Check if Multicall3 is supported on this chain
+  if (
+    !MULTICALL_CONTRACT_BY_CHAINID[
+      chainId as keyof typeof MULTICALL_CONTRACT_BY_CHAINID
+    ]
+  ) {
+    // Fallback to individual balance calls when Multicall3 is not supported
+    return await getTokenBalancesFallback(
+      normalizedTokenAddresses,
+      userAddresses,
+      provider,
+      includeNative,
+      20,
+    );
+  }
+
+  try {
+    // Create ERC20 balance calls
+    const erc20CallData = createERC20BalanceCalls(
+      normalizedTokenAddresses,
+      userAddresses,
+      provider,
+    );
+
+    let allCalls = erc20CallData.calls;
+    let allCallMapping = erc20CallData.callMapping;
+
+    // Add native token balance calls if requested
+    if (includeNative) {
+      const nativeCallData = createNativeTokenBalanceCalls(
+        userAddresses,
+        chainId,
+        provider,
+      );
+      allCalls = [...allCalls, ...nativeCallData.calls];
+      allCallMapping = [...allCallMapping, ...nativeCallData.callMapping];
+    }
+
+    // Execute all calls in batches
+    const maxCallsPerBatch = 300; // Limit calls per batch to avoid gas/size limits
+    const allResults: Aggregate3Result[] = [];
+
+    await reduceInBatchesSerially<Aggregate3Call, void>({
+      values: allCalls,
+      batchSize: maxCallsPerBatch,
+      initialResult: undefined,
+      eachBatch: async (_, batch) => {
+        const batchResults = await aggregate3(batch, chainId, provider);
+        allResults.push(...batchResults);
```

Comment:
optional nit - the `reduceInBatchesSerially` could return the array of results?
But TBH I'm fine with just pushing the results into this array.